### ISSUE TITLE
Make MachineDeployments version agnostic

### ISSUE THREAD

**Issue by furkatgofurov7** [kubernetes-sigs/cluster-api#6304]
**User Story**

As an operator, I would like to have a  version agnostic machinedeployment  so that whenever there is an upgrade it would not cause a rollout upgrade on existing worker nodes and keep them as is.

**Detailed Description**
We have a cluster with an N number of workers. If we want to scale out and do an upgrade (changing replicas and desired k8s version in MD), with the current behavior of MD, it ends up triggering a rollout upgrade resulting in re-provisioning of all the current nodes (deleting the current and creating new ones) which might be an expensive operation in some environments. 
What would be nice to have instead is, whenever we scale-out and do upgrade, there would be (or maybe there is already) a possibility of not touching the existing worker nodes at all (meaning, upgrade on all existing nodes can be handled by the user separately, ie by ssh ing into them and doing a manual upgrade), while all the newly created worker nodes would be using the desired k8s version. 
 
**Anything else you would like to add:**
One possible way could be using the `KubernetesVersion` field in the `ClusterConfiguration` configuration object which is valid for control plane nodes, but maybe we could make it valid for worker nodes as well (not sure if it is a problem) and use  `KubeadmConfigSpec.ClusterConfiguration.KubernetesVersion` field so that whenever we change k8s version in MD, it would not affect at all existing worker nodes (trigger re-provisioning), but desired k8s version is copied to KubeadmConfigTemplate and new worker nodes would be created using that version. I do not know much if that would cause any problems or break some existing behavior, so I would like to hear your opinions/suggestions/feedback. 

/kind feature

**Comment by vincepri** [comment]
@furkatgofurov7 This sound like in line with in place upgrade that we've talked about in the past. The biggest challenge to achieve what's described above is how Cluster API operates on the node itself without the use of SSH, but rather using an authenticated and authorized agent of some sorts.

The above suggestion of ignoring the version field is a major rewrite of the current logic which uses a hash of the the MachineTemplate object (which includes `spec.version`). A rewrite of the MachineDeployment logic is long overdue, and it should probably be based on what KCP does today, which is semantic comparison of field by field.

/milestone Next

**Comment by enxebre** [comment]
@furkatgofurov7 I definitely agree recreation rolling upgrades are not suitable for all scenarios, particularly true in bare metal environments.

Some context and possibilities available to enable this today:
>What would be nice to have instead is, whenever we scale-out and do upgrade, there would be (or maybe there is already) a possibility of not touching the existing worker nodes at all (meaning, upgrade on all existing nodes can be handled by the user separately, ie by ssh ing into them and doing a manual upgrade), while all the newly created worker nodes would be using the desired k8s version.

You can pretty much achieve the above by renouncing to some of the MachineDeployment opinionated automation and consuming MachineSets directly to have less automation and more control and see changes only apply to upcoming Machines.

Technically you could also just ignore machineDeployment version as the source truth and have your own out of band system managing the OS/Kube version transparently. This is not supported nor tested.

>The biggest challenge to achieve what's described above is how Cluster API operates on the node itself without the use of SSH, but rather using an authenticated and authorized agent of some sorts.

A possible way to solve this would be having a component (in-place upgrade controller) being authoritative source of truth for in-place upgrades/version per cluster with permissions to run pods in the Nodes, then this pods would be responsible for running the logic to operate the in-place upgrade and signal back to the authoritative source of truth.

I think this is common use case for bare metal envs and I would like CAPI to be able to offer some level of support / API contracts, e.g MachineDeployment upgradeType could be "replace" vs "in-place" and let "in-place" enable the workflow above.

This is though non trivial and It would require a thorough proposal in case there's alignment on the desire and scope. Happy to hear other thoughts.

**Comment by furkatgofurov7** [comment]
@vincepri @enxebre thanks for sharing your thoughts folks


> This sound like in line with in place upgrade that we've talked about in the past.

Hmm, most probably I am missing on these and would really appreciate it if you could share where these discussions happened.



> You can pretty much achieve the above by renouncing to some of the MachineDeployment opinionated automation and consuming MachineSets directly to have less automation and more control and see changes only apply to upcoming Machines.
> 
> Technically you could also just ignore machineDeployment version as the source truth and have your own out of band system managing the OS/Kube version transparently. This is not supported nor tested.

@enxebre thanks for the ideas, however, this would require pretty much user intervention and can lead to broken clusters when there is not enough care put into this workflow, since it is not supported by CAPI or tested as you already mentioned.



> A possible way to solve this would be having a component (in-place upgrade controller) being authoritative source of truth for in-place upgrades/version per cluster with permissions to run pods in the Nodes, then this pods would be responsible for running the logic to operate the in-place upgrade and signal back to the authoritative source of truth.
> 
> I think this is common use case for bare metal envs and I would like CAPI to be able to offer some level of support / API contracts, e.g MachineDeployment upgradeType could be "replace" vs "in-place" and let "in-place" enable the workflow above.

This sounds interesting, this would be indeed a desired capability that we could make use of in specific use cases, what we were thinking initially was having a "non-rollout-upgrade" kind of strategy added, but since MD uses hashes as stated, this might not be a feasible way to go.

**Comment by CecileRobertMichon** [comment]
Does `onDelete` rollout strategy help with this use case at all?

**Comment by furkatgofurov7** [comment]
> Does `onDelete` rollout strategy help with this use case at all?

@CecileRobertMichon  `onDelete` rollout strategy would not help, since it will trigger the deletion of machines in the old machine sets and then proceed to provision new ones, whereas in our use case we would like to keep the old machines as is (handle manual upgrade on them) to avoid going thorough reprovisioning cycle again.

**Comment by CecileRobertMichon** [comment]
That's not my understanding of "onDelete" from https://github.com/kubernetes-sigs/cluster-api/pull/4346

> Users control upgrading to a new MachineSet Configuration by triggering deletion of machines in the old machine set(s) with kubectl delete machine MACHINENAME. The machineDeployment controller waits for the machine to get deleted completely and then will proceed to provision the new replica with the new configuration.

> In this model: in order to complete the rollout the user is responsible for ultimately "replacing" every machine using old configuration. This allows the user to identify and execute any order of machine upgrades they choose. They control the velocity of the rollout as well by how quickly they remove machines in the old configuration. The rollout of a new machine does not occur until the old machine is fully removed in the current implementation.

**Comment by furkatgofurov7** [comment]
> Users control upgrading to a new MachineSet Configuration by triggering deletion of machines in the old machine set(s) with kubectl delete machine MACHINENAME. The machineDeployment controller waits for the machine to get deleted completely

@CecileRobertMichon yes, to use it, you have to manually delete existing machine right, so once the machine is deleted, the controller detects that the older replica set has N-1 replicas and spins up a new replica using the new machineSet.

**Comment by furkatgofurov7** [comment]
Unless I am not fully seeing the whole picture for  `onDelete`  rollout strategy usage, my main concern is how do we ensure the existing worker machines are left intact when `onDelete`  rollout strategy is used

**Comment by CecileRobertMichon** [comment]
> how do we ensure the existing worker machines are left intact when onDelete rollout strategy is used

If you don't delete the new machines they won't be touched. Only new machines will come up with the upgrades version. The existing machine won't be replaced until they are deleted.

**Comment by furkatgofurov7** [comment]
> If you don't delete the new machines they won't be touched

Is it even possible, as per original issue on that feature https://github.com/kubernetes-sigs/cluster-api/issues/4344#issue-836968632:

`It is a similar idea to the OnDelete strategy allowed in DaemonSet's:` https://kubernetes.io/docs/tasks/manage-daemon/update-daemon-set/ which states:

`OnDelete: With OnDelete update strategy, after you update a DaemonSet template, new DaemonSet pods will only be created when you manually delete old DaemonSet pods.`

However, I can verify it manually of course :slightly_smiling_face:

**Comment by CecileRobertMichon** [comment]
> new DaemonSet pods will only be created when you manually delete old DaemonSet pods

that's the key part, the controller won't delete and replace machines for you to rollout the machines, _only if the user manually deletes machine_ will new machines with the new "model" be created, which means in the meantime if you don't need to replace machines you can keep your old machines around forever. This doesn't solve in place upgrade per se but it does help with the concept of "manual rollout" so an automatic rollout doesn't get triggered when you update the machine deployment. 

but yes, please do verify manually so we're on the same page about what this does and doesn't do.

**Comment by furkatgofurov7** [comment]
@CecileRobertMichon sorry for the delay, I tested the "OnDelete" rollout strategy and realized that in fact, it covers the use case we have been discussing in this issue. Thanks a lot for the heads up!  
For anyone that might check on this issue in the future, this is how we achieved it by setting MD configs as:

```
 strategy:
    rollingUpdate:
      maxSurge: 0
      maxUnavailable: 1 
    type: OnDelete
```
so that when we scale out and update the spec of workers, MD keeps creating new workers without touching old ones, and once that is finished, and to stretch the case we had a bit more, we manually switched the strategy type back to `RollingUpdate`  which in turn automatically took care of old nodes (by deleting them all and recreating new ones with the new spec) so at the end, all worker nodes were at the same level. 

Also, really appreciate your input and sharing your thoughts on this issue @enxebre @vincepri :+1: 
I will go on and close this ticket. 

/close

**Comment by 24sama** [comment]
Hi, @furkatgofurov7. Is there any way to handle the KCP use case? I didn't find the `onDelete` strategy in the KCP.

**Comment by furkatgofurov7** [comment]
> **Anything else you would like to add:**
> One possible way could be using the `KubernetesVersion` field in the `ClusterConfiguration` configuration object which is valid for control plane nodes, but maybe we could make it valid for worker nodes as well (not sure if it is a problem) and use `KubeadmConfigSpec.ClusterConfiguration.KubernetesVersion` field so that whenever we change k8s version in MD, it would not affect at all existing worker nodes (trigger re-provisioning), but desired k8s version is copied to KubeadmConfigTemplate and new worker nodes would be created using that version.



> Hi, @furkatgofurov7. Is there any way to handle the KCP use case? I didn't find the `onDelete` strategy in the KCP.

@24sama hi! Would the above fulfill your requirements, although I have not tested it myself and can't confirm it?
Also, since this issue is already closed and was only specific to MD, I would suggest opening a new issue and referencing back to this one to have a better background on the new issue.
